hail/python/hail/matrixtable.py

```diff
@@ -2807,25 +2807,205 @@ def write(
+                if block_width > width:
+                    display_block_slices.append(slice(start_index, end_index))
+                    start_index = end_index
+                    block_width = display_column_widths[end_index] + 4
+                end_index += 1
+            display_block_slices.append(slice(start_index, end_index))
+
+            def format_hline(widths):
+                if not widths:
+                    return "++\n"
+                return '+-' + '-+-'.join(['-' * width for width in widths]) + '-+\n'
+
+            def pad(value: str, width: int, right_align: bool):
+                extra_count = width - len(value)
+                if right_align:
+                    return ' ' * extra_count + value
+                else:
+                    return value + ' ' * extra_count
+
+            def format_line(values: List[str], widths: List[int], right_align: List[bool]):
+                if not values:
+                    return "||\n"
+                values = map(pad, values, widths, right_align)
+                return '| ' + ' | '.join(values) + ' |\n'
+
+            ascii_str = ''
+            is_first_display_block = True
+            for block_slice in display_block_slices:
+                if is_first_display_block:
+                    is_first_display_block = False
+                else:
+                    ascii_str += '\n'
+
+                block_display_column_widths = display_column_widths[block_slice]
+                block_right_align = right_align[block_slice]
+                hline = format_hline(block_display_column_widths)
+
+                ascii_str += hline
+                ascii_str += format_line(
+                    display_column_names[block_slice], block_display_column_widths, block_right_align
+                )
+                ascii_str += hline
+                if should_show_types:
+                    ascii_str += format_line(
+                        truncated_types[block_slice], block_display_column_widths, block_right_align
+                    )
+                    ascii_str += hline
+                for row in truncated_values:
+                    ascii_str += format_line(row[block_slice], block_display_column_widths, block_right_align)
+                ascii_str += hline
+
+            if n_rows < matrix_table.count_rows():
+                row_count = len(rows)
+                ascii_str += f"showing top { row_count } { 'row' if row_count == 1 else 'rows' }\n"
+            total_n_cols = matrix_table.count_cols()
+            if n_cols != total_n_cols:
+                ascii_str += f"showing the first { n_cols } of { total_n_cols } columns"
+
+            return ascii_str
 
         def __repr__(self):
             return self.__str__()
 
         def _repr_html_(self):
-            s = self.table_show._repr_html_()
-            if self.displayed_n_cols != self.actual_n_cols:
+            # This method is not thoroughly tested and should be used carefully.
```

Comment:
We prefer to not include comments in the source code unless they're linking to a document that describes some jarringly abnormal behavior (like a known WONTFIX bug in a third-party library).

The HTML show methods have never been well-tested and that's OK for now.

Reply:
I will remove most of the comments in the next set of changes, but I would like to leave this one in.
```
# Here we must distinguish between matrix table columns
# and the columns that are used to display this matrix table.
 # Hail displays both the row fields and the matrix table's columns
# and entry fields as columns in a tabular format.
# We refer to the columns used to display the matrix table as "display columns".
```
I think it clarifies concepts in a way that would be difficult to understand from the code alone.